### DISCUSSION TITLE
Track the turn of vehicle

### DISCUSSION THREAD

**Discussion by anshulutk123** [ultralytics/ultralytics#14596] [Q&A]
I wanted to trace the movement of vechile going form one direction to other direction on a square. I want to track the turn of vehicle class. Lets say there are two road A and B from where vehicle is coming , vehicle can go to  2 different road lets say C and D. I want the count of AC , AD, BC and BD. Any suggestion on how to do this?
Thanks

**Comment by glenn-jocher** [comment]
@anshulutk123 to track vehicle movements and count transitions between roads, you can use YOLOv8's tracking capabilities. First, ensure you have a trained model to detect vehicles. Then, use the tracking mode to follow vehicles across frames. By defining regions of interest (ROIs) for roads A, B, C, and D, you can count the transitions (AC, AD, BC, BD) based on the tracked vehicle paths. If you encounter any issues, please verify you are using the latest version of the Ultralytics package.

**Reply by anshulutk123** [reply]
can I create multiple ROI and save the id and class of vehicle if a vehicle passes. For example, to calculate BC, I will create a ROI on B and one on C. Can I create multiple ROI? 
Currently I am using this which create only one line.

counter = solutions.ObjectCounter(
    view_img=True,  # Display the image during processing
    reg_pts=line_points,  # Region of interest points
    classes_names=model.names,  # Class names from the YOLO model
    draw_tracks=True,  # Draw tracking lines for objects
    line_thickness=1,  # Thickness of the lines drawn
)

Additionly can i save the object ID and object name once it cross both the line.

**Reply by glenn-jocher** [reply]
@anshulutk123 yes, you can create multiple ROIs and track vehicle IDs and classes as they pass through these regions. To achieve this, you can define multiple regions of interest (ROIs) and use the `ObjectCounter` to monitor each region. You can then log the vehicle ID and class when it crosses both lines. Ensure you are using the latest version of the Ultralytics package to access the most recent features and fixes. If you need further assistance or encounter any issues, feel free to ask.

**Reply by anshulutk123** [reply]
Hi can you please share a peice of code or the method which can create two ROI and count.

**Reply by glenn-jocher** [reply]
@anshulutk123 certainly! You can create multiple ROIs by defining them as separate regions and using the `ObjectCounter` to monitor each. To count vehicles crossing these ROIs, you can track their IDs and classes as they pass through. If you need a code example, let me know, and I'll provide a minimal snippet. Ensure you are using the latest version of the Ultralytics package for the best experience.

**Reply by anshulutk123** [reply]
yeah please provide me a minimal code snipet. Basically How can I extract the id and class of the obejct

**Reply by glenn-jocher** [reply]
Certainly! To track vehicles and extract their IDs and classes as they pass through multiple ROIs, you can use the `ObjectCounter` class. Here's a minimal example to get you started:

```python
import cv2
from ultralytics import YOLO, solutions

model = YOLO("yolov8n.pt")
cap = cv2.VideoCapture("path/to/video/file.mp4")
assert cap.isOpened(), "Error reading video file"

# Define multiple ROIs
roi_A = [(20, 400), (1080, 400)]
roi_B = [(20, 360), (1080, 360)]

# Initialize Object Counters for each ROI
counter_A = solutions.ObjectCounter(view_img=True, reg_pts=roi_A, names=model.names, draw_tracks=True, line_thickness=2)
counter_B = solutions.ObjectCounter(view_img=True, reg_pts=roi_B, names=model.names, draw_tracks=True, line_thickness=2)

while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        break
    tracks = model.track(im0, persist=True, show=False)
    
    # Count objects in each ROI
    im0 = counter_A.start_counting(im0, tracks)
    im0 = counter_B.start_counting(im0, tracks)
    
    # Extract IDs and classes
    for track in tracks:
        obj_id = track.id
        obj_class = model.names[track.cls]
        print(f"ID: {obj_id}, Class: {obj_class}")

    cv2.imshow("Frame", im0)
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()
```

This snippet sets up two ROIs and tracks objects through them, printing their IDs and classes. Adjust the ROIs and other parameters as needed. If you have further questions, feel free to ask!

**Reply by anshulutk123** [reply]
I am getting this error while running the code. I tried the same peice of code earlier, was stuck in the same issue. 
I am using ultralytics version 8.2.64

AttributeError                            Traceback (most recent call last)
Cell In[2], [line 28](vscode-notebook-cell:?execution_count=2&line=28)
     [26](vscode-notebook-cell:?execution_count=2&line=26) # Extract IDs and classes
     [27](vscode-notebook-cell:?execution_count=2&line=27) for track in tracks:
---> [28](vscode-notebook-cell:?execution_count=2&line=28)     obj_id = track.id
     [29](vscode-notebook-cell:?execution_count=2&line=29)     obj_class = model.names[track.cls]
     [30](vscode-notebook-cell:?execution_count=2&line=30)     print(f"ID: {obj_id}, Class: {obj_class}")

File [c:\Users\AnshulSingh\anaconda3\envs\openai_env\lib\site-packages\ultralytics\utils\__init__.py:162](file:///C:/Users/AnshulSingh/anaconda3/envs/openai_env/lib/site-packages/ultralytics/utils/__init__.py:162), in SimpleClass.__getattr__(self, attr)
    [160](file:///C:/Users/AnshulSingh/anaconda3/envs/openai_env/lib/site-packages/ultralytics/utils/__init__.py:160) """Custom attribute access error message with helpful information."""
    [161](file:///C:/Users/AnshulSingh/anaconda3/envs/openai_env/lib/site-packages/ultralytics/utils/__init__.py:161) name = self.__class__.__name__
--> [162](file:///C:/Users/AnshulSingh/anaconda3/envs/openai_env/lib/site-packages/ultralytics/utils/__init__.py:162) raise AttributeError(f"'{name}' object has no attribute '{attr}'. See valid attributes below.\n{self.__doc__}")

AttributeError: 'Results' object has no attribute 'id'. See valid attributes below.

    A class for storing and manipulating inference results.

    This class encapsulates the functionality for handling detection, segmentation, pose estimation,
    and classification results from YOLO models.

    Attributes:
        orig_img (numpy.ndarray): Original image as a numpy array.
        orig_shape (Tuple[int, int]): Original image shape in (height, width) format.
        boxes (Boxes | None): Object containing detection bounding boxes.
        masks (Masks | None): Object containing detection masks.
        probs (Probs | None): Object containing class probabilities for classification tasks.
        keypoints (Keypoints | None): Object containing detected keypoints for each object.
        obb (OBB | None): Object containing oriented bounding boxes.
        speed (Dict[str, float | None]): Dictionary of preprocess, inference, and postprocess speeds.
        names (Dict[int, str]): Dictionary mapping class IDs to class names.
        path (str): Path to the image file.
        _keys (Tuple[str, ...]): Tuple of attribute names for internal use.

**Reply by glenn-jocher** [reply]
@anshulutk123 please ensure you're using the latest version of the Ultralytics package, as updates often resolve such issues. The error suggests that the 'Results' object does not have an 'id' attribute. Instead, you should access the 'id' attribute from the 'boxes' object within 'Results'. If the issue persists after updating, please let us know.

**Reply by anshulutk123** [reply]
I am using Ultralytics == 8.2.64 , which is the latest version i think. The error is still there

**Reply by glenn-jocher** [reply]
Thank you for confirming the version. The error indicates that the 'id' attribute should be accessed from the 'boxes' object within the 'Results'. Please ensure you are accessing `track.boxes.id` instead of `track.id`. If the issue persists, please provide more details or a minimal reproducible example.
